core/src/ru/spbau/blackout/network/AndroidClient.java

```diff
@@ -0,0 +1,156 @@
+                                isInterrupted = true;
+                            }
+                        }
+                        out.writeBoolean(!isInterrupted);
+                        out.flush();
+                        break;
+                    default:
+                        break;
+                }
+            } while (gameState == GameState.WAITING && !isInterrupted);
+
+            final Thread outputToServerThread = new Thread(() -> {
+                while (!isInterrupted) {
+                    try {
+                        Vector2 sending = new Vector2(velocityToSend);
+                        out.writeObject(sending);
+                        out.flush();
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        isInterrupted = true;
+                    }
+                    // sleeping to not send position too often.
+                    try {
+                        sleep(Network.SLEEPING_TIME_TO_ACHIEVE_FRAME_RATE);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                        isInterrupted = true;
+                    }
+                }
+            });
+            outputToServerThread.start();
+
+            // waiting for the first world to arrive
+            socket.setSoTimeout(0);
+            while (!isInterrupted) {
+                // should read game world here from inputStream
+                final GameWorld currentWorld = gameScreen.getGameWorld();
+
+                byte[] serializedWorld = (byte[]) in.readObject();
+
+                Gdx.app.log("ANDROID", "TRYING TO ACQUIRE LOCK");
+                //noinspection SynchronizationOnLocalVariableOrMethodParameter
+                synchronized (currentWorld) {
+                    Gdx.app.log("ANDROID", "STARTING DESERIALIZATION ON IT");
+                    currentWorld.inplaceDeserialize(new ObjectInputStream(new ByteArrayInputStream(serializedWorld)));
+                    Gdx.app.log("ANDROID", "FINISHING DESERIALIZATION ON IT");
+                }
+                Gdx.app.log("ANDROID", "RELEASING TO ACQUIRE LOCK");
+
+                // should get worlds regularly after the first one
+                socket.setSoTimeout(Network.SOCKET_IO_TIMEOUT_MS);
+            }
+        } catch (UnknownHostException e) {
+            Gdx.app.log(TAG, "Don't know about host " + HOST_NAME);
+        } catch (ClassNotFoundException | IOException e) {
+            e.printStackTrace();
+        } finally {
+            isInterrupted = true;
+            Gdx.app.postRunnable(() -> {
+                final MenuScreen menuScreen = table.getMenuScreen();
+                menuScreen.changeMiddleTable(PlayScreenTable.getTable(menuScreen));
+                BlackoutGame.getInstance().getScreenManager().setScreen(menuScreen);
+            });
+        }
+    }
+
+    public void interrupt() {
+        isInterrupted = true;
```

Comment:
To avoid confusion with the default synchronization mechanisms, can you name this method something like "stop"? Right now this code does not wake up any internal thread that is sleep()ing or wait()ing. one way to do it properly is

public synchronized void stop() {
  shouldStop = true; // new isInterrupted
  notifyAll(); // to wake up internal threads and let them see the updated flag

  // and maybe even wait for the threads to actually stop
}